opal/mca/common/ofi/common_ofi.c

```diff
@@ -778,6 +780,168 @@ static uint32_t get_package_rank(opal_process_info_t *process_info)
+ * @param[in]   num_providers   number of providers
+ * @param[in]   device_rank     local rank on the accelerator
+ * @param[out]  provider        pointer to the selected provider
+ * @return      OPAL_SUCCESS if a provider is successfully selected
+ *              OPAL_ERR_NOT_AVAILABLE if a provider cannot be decided deterministically
+ *              OPAL_ERROR if a fatal error happened
+ */
+static int find_nearest_provider_from_accelerator(struct fi_info *provider_list,
+                                                  size_t num_providers, uint32_t device_rank,
+                                                  struct fi_info **provider)
+{
+    hwloc_obj_t accl_dev = NULL, prov_dev = NULL, common_ancestor = NULL;
+    int ret = -1, accl_id = -1, depth = -1, max_common_ancestor_depth = -1;
+    opal_accelerator_pci_attr_t accl_pci_attr = {0};
+    struct fi_info *current_provider = NULL;
+    struct fi_pci_attr pci = {0};
+    uint32_t near_provider_count = 0, provider_rank = 0;
+    uint32_t distances[num_providers], *distance = distances;
+
+    memset(distances, 0, sizeof(distances));
+
+    ret = opal_accelerator.get_device(&accl_id);
+    if (OPAL_SUCCESS != ret) {
+        opal_output_verbose(1, opal_common_ofi.output, "%s:%d:Accelerator is not available",
+                            __FILE__, __LINE__);
+        return OPAL_ERR_NOT_AVAILABLE;
+    }
+
+    ret = opal_accelerator.get_device_pci_attr(accl_id, &accl_pci_attr);
+    if (OPAL_SUCCESS != ret) {
+        opal_output_verbose(1, opal_common_ofi.output,
+                            "%s:%d:Accelerator PCI info is not available", __FILE__, __LINE__);
+        return OPAL_ERROR;
+    }
+
+    accl_dev = hwloc_get_pcidev_by_busid(opal_hwloc_topology, accl_pci_attr.domain_id,
+                                         accl_pci_attr.bus_id, accl_pci_attr.device_id,
+                                         accl_pci_attr.function_id);
+    if (NULL == accl_dev) {
+        opal_output_verbose(1, opal_common_ofi.output,
+                            "%s:%d:Failed to find accelerator PCI device", __FILE__, __LINE__);
+        return OPAL_ERROR;
+    }
+
+    opal_output_verbose(1, opal_common_ofi.output,
+                        "%s:%d:Found accelerator device %d: %04x:%02x:%02x.%x VID: %x DID: %x",
+                        __FILE__, __LINE__, accl_id, accl_pci_attr.domain_id, accl_pci_attr.bus_id,
+                        accl_pci_attr.device_id, accl_pci_attr.function_id,
+                        accl_dev->attr->pcidev.vendor_id, accl_dev->attr->pcidev.device_id);
+
+    current_provider = provider_list;
+    while (NULL != current_provider) {
+        common_ancestor = NULL;
+        if (0 == check_provider_attr(provider_list, current_provider)
+            && (NULL != current_provider->nic && NULL != current_provider->nic->bus_attr
+                && current_provider->nic->bus_attr->bus_type == FI_BUS_PCI)) {
+            pci = current_provider->nic->bus_attr->attr.pci;
+
```

Comment:
Just a suggestion: I think abstracting out retrieving the ```pci``` into an inline function would allow to remove multiple conditions of the if statement in lines 873 and 874 and would improve the readability a bit by moving those checks into the new function, e.g.

```
if (0 == check_provider_attr(provider_list, current_provider) {
   ret = get_pci_attr(current_provider, &pci);
   if ( OPAL_SUCCESS != ret) {
        continue;
   }
   ...
```
(or something like that)

Reply:
I see. Good catch. I will refactor similar logic in an update. Thanks.

Reply:
@edgargabriel I forgot to mention but do you have a test case for rocm on hand to verify the impl?

I only tested it for cuda + EFA. The idea is to verify the NIC -> GPU pairing is expected, i.e. match lspci

FYR I ran this
```
/home/ec2-user/ompi/install/bin/mpirun -mca mtl ofi --wdir . --hostfile hostfile -n 36 --map-by ppr:36:node --rank-by slot --bind-to core -x LD_LIBRARY_PATH=/home/ec2-user/libfabric/install/lib:$LD_LIBRARY_PATH -x FI_EFA_USE_DEVICE_RDMA=1 -x FI_EFA_ENABLE_SHM_TRANSFER=0 -x PATH=/home/ec2-user/ompi/install/bin:$PATH -x OMPI_MCA_mtl_ofi_verbose=1  --tag-output --report-bindings /home/ec2-user/osu-micro-benchmarks/install/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_mbw_mr -m 4
...
[1,8]<stderr>: [queue-c6inmetal-st-c6inmetal-1:04941] Rank 8 bound to package[0][core:8]
...
[1,8]<stderr>: [queue-c6inmetal-st-c6inmetal-1:04971] mtl_ofi_component.c:360: mtl:ofi:provider:domain: rdmap7s0-rdm
...
```

Reply:
not sure, let me think about it. I would have to setup a testcase where we use multiple procs per node, but the procs on the node cannot communicate GPU buffers amongst each other, since ompi currently can only select a single provider, i.e. we cannot do both libfabrics verbs and sm  (with ROCr support) at the same . So it would have to be carefully tuned scenario.

Reply:
I see. Thanks for letting me know. I will trust your judgement then. Just want to get a sanity check.

Reply:
Refactored the nic check code in update.